src/librustc_target/abi/call/riscv.rs

```diff
@@ -1,49 +1,341 @@
 // Reference: RISC-V ELF psABI specification
 // https://github.com/riscv/riscv-elf-psabi-doc
+//
+// Reference: Clang RISC-V ELF psABI lowering code
+// https://github.com/llvm/llvm-project/blob/8e780252a7284be45cf1ba224cabd884847e8e92/clang/lib/CodeGen/TargetInfo.cpp#L9311-L9773
 
-use crate::abi::call::{ArgAbi, FnAbi};
+use crate::abi::call::{ArgAbi, ArgAttribute, FnAbi, PassMode, Reg, Uniform};
+use crate::abi::{self, Abi, FieldPlacement, LayoutOf, TyLayout, TyLayoutMethods};
+use crate::spec::HasTargetSpec;
+
+#[derive(Copy, Clone)]
+enum FieldKind {
+    Float,
+    Integer,
+    Unknown,
+}
+
+#[derive(Copy, Clone)]
+enum FloatArgPass {
+    FloatPair,
+    Float,
+    MixedPair,
+}
+
+fn is_vector<'a, Ty>(arg: &ArgAbi<'a, Ty>) -> bool {
+    match arg.layout.abi {
+        Abi::Vector { .. } => true,
+        _ => false,
+    }
+}
+
+fn is_aggregate<'a, Ty>(arg: &ArgAbi<'a, Ty>) -> bool {
```

Comment:
This should probably named `is_aggregate_or_vector` or `is_riscv_aggregate`, to make it harder to configure with `.is_aggregate()`.